File: openbio_singlecell/monocle2.py
"""File exchange for native Monocle 2; Python retains ownership of AnnData."""

from __future__ import annotations

import hashlib
import inspect
import tempfile
from pathlib import Path

from . import expression_source, r_runtime
from .expression_source import ExpressionSourceSpec
from .r_runtime import run_r_script
from .worker_protocol import read_json, write_json

MONOCLE2_KIND = "OPENBIO_MONOCLE2_CDS"
MONOCLE2_CODEC = "monocle2-cds-rds-v1"
MONOCLE2_PACKAGES = ("monocle", "DDRTree", "Matrix", "jsonlite", "Biobase", "VGAM", "igraph", "dplyr", "ggplot2")
MONOCLE2_SCRIPT = Path(__file__).with_name("r") / "monocle2.R"
MONOCLE2_COMPAT_SCRIPT = MONOCLE2_SCRIPT.with_name("monocle2_compat.R")
MONOCLE2_SOURCE = ExpressionSourceSpec(description="Monocle 2 expression source", include_raw=True)


def _hash_file(path):
    digest = hashlib.sha256()
    with Path(path).open("rb") as stream:
        for chunk in iter(lambda: stream.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def _frame_payload(frame):
    import numpy as np
    import pandas as pd

    def scalar(value):
        if pd.isna(value):
            return None
        if isinstance(value, (float, np.floating)) and np.isinf(value):
            return "Inf" if value > 0 else "-Inf"
        return value.item() if isinstance(value, np.generic) else value

    columns = {}
    for name in frame.columns:
        column = frame[name]
        if isinstance(column.dtype, pd.CategoricalDtype):
            record = {"type": "category", "levels": [scalar(value) for value in column.cat.categories],
                      "ordered": bool(column.cat.ordered)}
        elif pd.api.types.is_bool_dtype(column.dtype):
            record = {"type": "boolean"}
        elif pd.api.types.is_integer_dtype(column.dtype):
            record = {"type": "integer"}
        elif pd.api.types.is_numeric_dtype(column.dtype):
            record = {"type": "number"}
        else:
            record = {"type": "string"}
        record["values"] = [scalar(value) for value in column]
        columns[str(name)] = record
    return {"index": frame.index.tolist(), "columns": columns}


def _read_frame(path):
    import pandas as pd

    payload = read_json(path)
    data, aliases = {}, {}
    dtypes = {"string": "string", "integer": "Int64", "number": "Float64", "boolean": "boolean"}
    for name, column in payload["columns"].items():
        if column["type"] == "category":
            data[name] = pd.Categorical(column["values"], categories=column["levels"], ordered=column["ordered"])
        else:
            data[name] = pd.array(column["values"], dtype=dtypes[column["type"]])
        if "original_name" in column:
            aliases[name] = column["original_name"]
    frame = pd.DataFrame(data, index=pd.Index(payload["index"], dtype="object"))
    frame.attrs["r_column_aliases"] = aliases
    return frame


def read_cds(root):
    root = Path(root).resolve(strict=True)
    metadata = read_json(root / "metadata.json")
    if not isinstance(metadata, dict) or metadata.get("kind") != MONOCLE2_KIND or metadata.get("codec") != MONOCLE2_CODEC:
        raise ValueError("Expected a Monocle 2 CellDataSet artifact.")
    if metadata.get("schema_version") != 1 or not isinstance(metadata.get("files"), dict):
        raise ValueError("Monocle 2 artifact metadata is invalid.")
    files = metadata["files"]
    if not {"state.rds", "cell_metadata.json", "ordering_genes.json", "summary.json"}.issubset(files):
        raise ValueError("Monocle 2 artifact is missing native state or result sidecars.")
    for name, digest in files.items():
        path = root / name
        if Path(name).name != name or path.resolve().parent != root or _hash_file(path) != digest:
            raise ValueError(f"Monocle 2 artifact file does not match its recorded content: {name!r}")
    return metadata


def _seal_cds(root, summary, runtime, source):
    names = ["state.rds", "cell_metadata.csv", "cell_metadata.json", "ordering_genes.csv", "ordering_genes.json", "summary.json"]
    if (root / "embedding.csv").exists():
        names.extend(["embedding.csv", "embedding.json"])
    metadata = {
        "schema_version": 1, "kind": MONOCLE2_KIND, "codec": MONOCLE2_CODEC,
        "files": {name: _hash_file(root / name) for name in names},
        "n_obs": summary["n_obs"], "n_vars": summary["n_vars"], "operation": summary["operation"],
        "versions": summary["versions"], "source": source,
        "runtime_version": runtime["r_version"],
    }
    write_json(root / "metadata.json", metadata)
    return metadata


def _attach_results(adata, cells, embedding, summary, parameters):
    import pandas as pd

    if not {"Pseudotime", "State"}.issubset(cells.columns):
        raise ValueError("Monocle 2 export requires orderCells results (State and Pseudotime).")
    cells = cells.set_index("cell_id")
    embedding = embedding.set_index("cell_id")
    for frame in (cells, embedding):
        if not adata.obs_names.is_unique or not frame.index.is_unique:
            raise ValueError("Monocle 2 attachment requires unique cell identifiers.")
        if len(frame) != adata.n_obs or len(frame.index.difference(adata.obs_names)):
            raise ValueError("Monocle 2 results must identify the same cells as the selected AnnData.")
    embedding_key = parameters["embedding_key"]
    pseudotime_key = parameters["pseudotime_key"]
    state_key = parameters["state_key"]
    if not all(isinstance(key, str) and key for key in (embedding_key, pseudotime_key, state_key)):
        raise ValueError("Monocle 2 output names cannot be empty.")
    if pseudotime_key == state_key:
        raise ValueError("Pseudotime and State require distinct observation columns.")
    if not parameters["overwrite_existing"]:
        existing = [key for key in (pseudotime_key, state_key) if key in adata.obs]
        if embedding_key in adata.obsm:
            existing.append(embedding_key)
        if existing:
            raise ValueError(f"Monocle 2 output fields already exist: {existing}; enable overwrite_existing to replace.")
    aligned = cells.reindex(adata.obs_names)
    adata.obs[pseudotime_key] = aligned["Pseudotime"].to_numpy(dtype=float)
    adata.obs[state_key] = pd.Categorical(aligned["State"], categories=list(summary["state_counts"]))
    adata.obsm[embedding_key] = embedding.reindex(adata.obs_names).to_numpy(dtype=float)
    return adata


def run_analysis(operation, runtime, output_dir, parameters, *, adata=None, cds_path=None, script_path=None):
    from scipy import sparse
    from scipy.io import mmwrite

    output_dir = Path(output_dir).resolve()
    parameters = dict(parameters)
    attachment = parameters if operation == "export" else None
    if attachment is not None:
        parameters = {}
    request = {"operation": operation, "output_dir": str(output_dir)}
    source = None
    with tempfile.TemporaryDirectory(prefix=".monocle-exchange-", dir=output_dir.parent) as exchange:
        exchange = Path(exchange)
        if operation == "create":
            selected = MONOCLE2_SOURCE.resolve(adata, parameters.pop("source", None))
            matrix = selected.matrix(adata)
            var = (adata.raw.var if selected.use_raw else adata.var).copy()
            symbol_column = parameters.pop("gene_short_name_column", "")
            var["gene_short_name"] = var[symbol_column] if symbol_column else var.index
            mmwrite(exchange / "expression.mtx", sparse.coo_matrix(matrix.T), precision=17, symmetry="general")
            write_json(exchange / "obs.json", _frame_payload(adata.obs))
            write_json(exchange / "var.json", _frame_payload(var))
            request.update(matrix_path=str(exchange / "expression.mtx"), obs_path=str(exchange / "obs.json"),
                           var_path=str(exchange / "var.json"))
            source = {"expression": selected.parameters(), "gene_short_name_column": symbol_column}
        else:
            metadata = read_cds(cds_path)
            source = metadata.get("source")
            request["input_rds"] = str(Path(cds_path).resolve() / "state.rds")
        request["parameters"] = parameters
        summary = run_r_script(runtime, Path(script_path) if script_path else MONOCLE2_SCRIPT, request, output_dir)
    result = {"summary": summary, "output_dir": output_dir}
    if (output_dir / "state.rds").exists():
        _seal_cds(output_dir, summary, runtime, source)
        result["cds_path"] = output_dir
    table_file = {
        "ordering_genes": "ordering_genes.json", "order_cells": "cell_metadata.json",
        "differential_test": "table.json", "beam": "table.json", "export": "cell_metadata.json",
    }.get(operation)
    if table_file:
        result["table"] = _read_frame(output_dir / table_file)
        if result["table"].attrs["r_column_aliases"]:
            summary["table_column_aliases"] = result["table"].attrs["r_column_aliases"]
    if (output_dir / "plot.png").exists():
        result["plot_path"] = output_dir / "plot.png"
    if attachment is not None:
        result["adata"] = _attach_results(
            adata, result["table"], _read_frame(output_dir / "embedding.json"), summary, attachment,
        )
    return result


_PORTABLE_HEADER = '''from __future__ import annotations
import hashlib
import json
import os
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, Literal, cast

def read_json(path):
    return json.loads(Path(path).read_text(encoding="utf-8"))

def write_json(path, value):
    Path(path).write_text(json.dumps(value, ensure_ascii=False, allow_nan=False), encoding="utf-8")

OUTPUT_TAIL_BYTES = 64 * 1024
ExpressionSourceKind = Literal["X", "raw", "layer"]
DynamicExpressionSource = Mapping[str, object]
'''


def _runtime_source():
    return "\n\n".join(inspect.getsource(function) for function in (
        r_runtime.RRuntimeError, r_runtime._child_environment, r_runtime._description_identity,
        r_runtime._invoke, r_runtime.validate_r_runtime, r_runtime.run_r_script,
    ))


def reproduction_code(operation, parameters, runtime):
    parts = [_PORTABLE_HEADER, _runtime_source(), inspect.getsource(expression_source.ExpressionSource),
             inspect.getsource(expression_source.ExpressionSourceSpec),
             f"MONOCLE2_KIND = {MONOCLE2_KIND!r}\nMONOCLE2_CODEC = {MONOCLE2_CODEC!r}\n"
             "MONOCLE2_SOURCE = ExpressionSourceSpec(description='Monocle 2 expression source', include_raw=True)\n"]
    parts.extend(inspect.getsource(function) for function in (
        _hash_file, _frame_payload, _read_frame, read_cds, _seal_cds, _attach_results, run_analysis,
    ))
    parts.append(f'''
def run_monocle2(adata=None, cds_path=None, *, output_dir, r_runtime=None):
    runtime = {runtime!r} if r_runtime is None else r_runtime
    if isinstance(runtime, str):
        runtime = json.loads(runtime)
    destination = Path(output_dir).resolve()
    destination.mkdir(parents=True, exist_ok=False)
    if {operation!r} == "export":
        adata = adata.copy()
    with tempfile.TemporaryDirectory(prefix="monocle2-source-") as bundle:
        script = Path(bundle) / "monocle2.R"
        script.write_text({MONOCLE2_SCRIPT.read_text(encoding="utf-8")!r}, encoding="utf-8")
        script.with_name("monocle2_compat.R").write_text(
            {MONOCLE2_COMPAT_SCRIPT.read_text(encoding="utf-8")!r}, encoding="utf-8")
        return run_analysis({operation!r}, runtime, destination, {parameters!r},
                            adata=adata, cds_path=cds_path, script_path=script)
''')
    return "\n\n".join(parts) + "\n"


def runtime_reproduction_code(parameters):
    return _PORTABLE_HEADER + "\n" + _runtime_source() + f'''
def probe_monocle2_runtime():
    options = {parameters!r}
    runtime = {{key: options[key] for key in ("r_home", "library_paths", "path_prefix")}}
    executable = shutil.which(options["rscript"], path=_child_environment(runtime).get("PATH"))
    if executable is None:
        raise ValueError("Rscript was not found in the selected environment.")
    executable = os.path.normcase(os.path.realpath(executable))
    stat = Path(executable).stat()
    runtime.update(executable=executable, size=stat.st_size, mtime_ns=stat.st_mtime_ns)
    with tempfile.TemporaryDirectory(prefix="monocle2-probe-") as bundle:
        root = Path(bundle)
        script = root / "probe.R"
        script.write_text({r_runtime.PROBE_SCRIPT.read_text(encoding="utf-8")!r}, encoding="utf-8")
        result = _invoke(runtime, script, {{"packages": {list(MONOCLE2_PACKAGES)!r}, "output_dir": bundle}}, root, timeout=60)
    for package in result["packages"].values():
        package["description"] = _description_identity(Path(package["path"]) / "DESCRIPTION")
    return {{**runtime, **result}}
'''

File: openbio_singlecell/r/monocle2.R
# One native Monocle2 operation per process. Python owns the complete AnnData.
suppressPackageStartupMessages(library(monocle))
script_path <- sub("^--file=", "", grep("^--file=", commandArgs(), value = TRUE)[[1]])
source(file.path(dirname(normalizePath(script_path)), "monocle2_compat.R"), local = TRUE)
backend <- monocle2_backend()

read_frame <- function(path) {
  metadata <- jsonlite::read_json(path, simplifyVector = FALSE)
  result <- data.frame(row.names = unlist(metadata$index, use.names = FALSE))
  for (name in names(metadata$columns)) {
    column <- metadata$columns[[name]]
    values <- unlist(lapply(column$values, function(value) {
      if (is.null(value)) NA else value
    }), use.names = FALSE)
    result[[name]] <- switch(column$type,
      category = if (is.null(column$levels)) factor(values, ordered = isTRUE(column$ordered)) else
        factor(values, levels = unlist(column$levels, use.names = FALSE),
               ordered = isTRUE(column$ordered)),
      integer = as.numeric(values),
      number = as.numeric(values),
      boolean = as.logical(values),
      string = as.character(values),
      stop(paste("Unknown metadata column type:", column$type)))
  }
  result
}

request <- jsonlite::read_json(commandArgs(trailingOnly = TRUE)[[1]], simplifyVector = TRUE)
parameters <- request$parameters
out <- function(name) file.path(request$output_dir, name)
write_frame <- function(frame, csv_name) {
  original_names <- colnames(frame)
  colnames(frame) <- make.unique(original_names)
  columns <- lapply(seq_along(frame), function(i) {
    column <- frame[[i]]
    type <- if (is.factor(column)) "category" else if (is.logical(column)) "boolean" else
      if (is.integer(column)) "integer" else if (is.numeric(column)) "number" else "string"
    values <- if (type %in% c("category", "string")) as.character(column) else column
    values <- unname(lapply(values, function(value) {
      if (is.na(value)) NULL else if (is.numeric(value) && is.infinite(value)) {
        if (value > 0) "Inf" else "-Inf"
      } else value
    }))
    record <- list(type = type, values = values)
    if (type == "category") {
      record$levels <- as.list(levels(column))
      record$ordered <- is.ordered(column)
    }
    if (colnames(frame)[i] != original_names[i]) record$original_name <- original_names[i]
    record
  })
  names(columns) <- colnames(frame)
  jsonlite::write_json(list(index = as.list(rownames(frame)), columns = columns),
    out(sub("\\.csv$", ".json", csv_name)), auto_unbox = TRUE, null = "null", na = "null", digits = NA)
  utils::write.csv(frame, out(csv_name), row.names = FALSE, na = "")
}
warnings <- character()
effective_parameters <- parameters
native_parameters <- function(exclude = character()) {
  arguments <- parameters[setdiff(names(parameters), c("extra_parameters", exclude))]
  if (!is.null(parameters$extra_parameters)) {
    arguments <- utils::modifyList(arguments, parameters$extra_parameters, keep.null = TRUE)
  }
  if (any(names(arguments) %in% c("cds", "cds_subset", "object"))) {
    stop("The input CellDataSet is supplied by the artifact, not by parameters.")
  }
  effective_parameters <<- arguments
  arguments
}
save_plot <- function(plot) {
  width <- if (is.null(parameters$width)) 8 else parameters$width
  height <- if (is.null(parameters$height)) 6 else parameters$height
  dpi <- if (is.null(parameters$dpi)) 150 else parameters$dpi
  grDevices::png(out("plot.png"), width = width, height = height, units = "in", res = dpi)
  tryCatch(print(plot), finally = grDevices::dev.off())
}

withCallingHandlers({
  cds <- switch(request$operation,
    create = {
      expression <- methods::as(Matrix::readMM(request$matrix_path), "CsparseMatrix")
      obs <- read_frame(request$obs_path)
      var <- read_frame(request$var_path)
      rownames(expression) <- rownames(var)
      colnames(expression) <- rownames(obs)
      family_name <- parameters$expression_family
      if (is.null(family_name)) family_name <- "negbinomial.size"
      families <- c("negbinomial.size", "negbinomial", "tobit", "gaussianff")
      if (!family_name %in% families) stop(paste("Unknown expression family:", family_name))
      limit <- parameters$lowerDetectionLimit
      if (is.null(limit)) limit <- 0.1
      cds <- monocle::newCellDataSet(expression,
        phenoData = methods::new("AnnotatedDataFrame", data = obs),
        featureData = methods::new("AnnotatedDataFrame", data = var),
        expressionFamily = getExportedValue("VGAM", family_name)(),
        lowerDetectionLimit = limit)
      if (isTRUE(parameters$estimate_size_factors)) {
        cds <- do.call(BiocGenerics::estimateSizeFactors,
                       c(list(object = cds), parameters$size_factor_parameters))
      }
      if (isTRUE(parameters$estimate_dispersions)) {
        cds <- do.call(backend$estimateDispersions,
                       c(list(object = cds), parameters$dispersion_parameters))
      }
      if (isTRUE(parameters$detect_genes)) {
        cds <- do.call(monocle::detectGenes, c(list(cds = cds), parameters$detect_parameters))
      }
      cds
    },
    ordering_genes = {
      cds <- readRDS(request$input_rds)
      genes <- switch(parameters$method,
        explicit = parameters$genes,
        var_column = {
          if (!parameters$var_column %in% colnames(Biobase::fData(cds))) {
            stop(paste("Feature annotation column not found:", parameters$var_column))
          }
          rownames(cds)[Biobase::fData(cds)[[parameters$var_column]] %in% TRUE]
        },
        dispersion = {
          dispersions <- monocle::dispersionTable(cds)
          mean_expression <- parameters$mean_expression
          if (is.null(mean_expression)) mean_expression <- 0.5
          dispersion_fold <- parameters$dispersion_fold
          if (is.null(dispersion_fold)) dispersion_fold <- 1
          dispersions$gene_id[dispersions$mean_expression >= mean_expression &
            dispersions$dispersion_empirical >= dispersion_fold * dispersions$dispersion_fit]
        },
        stop(paste("Unknown ordering gene method:", parameters$method)))
      monocle::setOrderingFilter(cds, genes)
    },
    ddrtree = {
      cds <- readRDS(request$input_rds)
      arguments <- native_parameters()
      if (!is.null(arguments$reduction_method) && !identical(arguments$reduction_method, "DDRTree")) {
        stop("Advanced parameters cannot change the DDRTree algorithm: reduction_method must be 'DDRTree'.")
      }
      had_ordering <- length(cds@dim_reduce_type) &&
        length(cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell) > 0
      # Native reduceDimension rebuilds graph projections but leaves old ordering auxiliaries behind.
      cds@auxOrderingData[["DDRTree"]] <- NULL
      cds <- do.call(monocle::reduceDimension, c(list(cds = cds), arguments))
      if (had_ordering) {
        Biobase::pData(cds)$State <- NULL
        Biobase::pData(cds)$Pseudotime <- NULL
        warning("The new DDRTree graph invalidated the previous ordering; State and Pseudotime were removed from this output. Run Order Cells again.")
      }
      cds
    },
    order_cells = {
      cds <- readRDS(request$input_rds)
      do.call(backend$orderCells, c(list(cds = cds), native_parameters()))
    },
    trajectory_plot = {
      cds <- readRDS(request$input_rds)
      plot <- do.call(backend$plot_cell_trajectory,
        c(list(cds = cds), native_parameters(c("width", "height", "dpi"))))
      save_plot(plot)
      cds
    },
    differential_test = {
      cds <- readRDS(request$input_rds)
      test_cds <- if (length(parameters$genes)) cds[parameters$genes, ] else cds
      result <- do.call(monocle::differentialGeneTest,
        c(list(cds = test_cds), native_parameters("genes")))
      result <- data.frame(gene_id = rownames(result), result, check.names = FALSE)
      write_frame(result, "table.csv")
      cds
    },
    gene_trends = {
      cds <- readRDS(request$input_rds)
      plot_cds <- if (length(parameters$genes)) cds[parameters$genes, ] else cds
      plot <- do.call(monocle::plot_genes_in_pseudotime,
        c(list(cds_subset = plot_cds), native_parameters(c("genes", "width", "height", "dpi"))))
      save_plot(plot)
      cds
    },
    beam = {
      cds <- readRDS(request$input_rds)
      test_cds <- if (length(parameters$genes)) cds[parameters$genes, ] else cds
      result <- do.call(backend$BEAM,
        c(list(cds = test_cds), native_parameters("genes")))
      result <- data.frame(gene_id = rownames(result), result, check.names = FALSE)
      write_frame(result, "table.csv")
      cds
    },
    export = {
      cds <- readRDS(request$input_rds)
      if (!length(monocle::reducedDimS(cds)) || !length(cds@dim_reduce_type) ||
          !length(cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell)) {
        stop("Run Monocle 2 Order Cells before export; native ordering results are not available in this CellDataSet.")
      }
      cds
    },
    stop(paste("Unknown Monocle2 operation:", request$operation)))

  if (request$operation %in% c("create", "ordering_genes", "ddrtree", "order_cells")) {
    saveRDS(cds, out("state.rds"))
  }
  cell_data <- data.frame(cell_id = colnames(cds), row.names = colnames(cds), check.names = FALSE)
  for (name in intersect(c("Size_Factor", "num_genes_expressed", "Pseudotime", "State"),
                         colnames(Biobase::pData(cds)))) {
    cell_data[[name]] <- Biobase::pData(cds)[[name]]
  }
  write_frame(cell_data, "cell_metadata.csv")
  gene_data <- data.frame(gene_id = rownames(cds), row.names = rownames(cds), check.names = FALSE)
  for (name in intersect(c("gene_short_name", "use_for_ordering", "num_cells_expressed"),
                         colnames(Biobase::fData(cds)))) {
    gene_data[[name]] <- Biobase::fData(cds)[[name]]
  }
  write_frame(gene_data, "ordering_genes.csv")
  if (length(monocle::reducedDimS(cds))) {
    embedding <- as.data.frame(t(monocle::reducedDimS(cds)))
    colnames(embedding) <- paste0("DDRTree", seq_len(ncol(embedding)))
    embedding <- data.frame(cell_id = rownames(embedding), embedding, check.names = FALSE)
    write_frame(embedding, "embedding.csv")
  }
  summary <- list(operation = request$operation, n_obs = ncol(cds), n_vars = nrow(cds),
                  versions = list(R = as.character(getRversion()),
                    monocle = as.character(utils::packageVersion("monocle")),
                    igraph = as.character(utils::packageVersion("igraph")),
                    dplyr = as.character(utils::packageVersion("dplyr")),
                    DDRTree = as.character(utils::packageVersion("DDRTree"))),
                  compatibility = backend$compatibility,
                  effective_parameters = effective_parameters,
                  ordering_gene_count = sum(Biobase::fData(cds)$use_for_ordering),
                  warnings = as.list(unique(warnings)))
  if ("State" %in% colnames(Biobase::pData(cds))) {
    summary$state_counts <- as.list(table(Biobase::pData(cds)$State))
  }
  if (length(cds@dim_reduce_type)) {
    summary$root_cell <- cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell
  }
  if (request$operation == "order_cells") {
    summary$root_method <- if (!is.null(effective_parameters$root_state)) "state" else
      if (isTRUE(effective_parameters$reverse)) "reverse" else "automatic"
  }
  jsonlite::write_json(summary, out("summary.json"), auto_unbox = TRUE,
                       null = "null", na = "null", digits = NA)
}, warning = function(warning) {
  warnings <<- c(warnings, conditionMessage(warning))
})
